src/edge/acrManager.ts

```diff
@@ -0,0 +1,224 @@
+        if (acrRepoItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        const acrTagItem: vscode.QuickPickItem = await this.selectAcrTag(registryUrl, acrRepoItem.label);
+        if (acrTagItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        this.loginToAcr(acrRegistryItem.registry, acrRegistryItem.azureSubscription);
+
+        return acrTagItem.description;
+    }
+
+    private async selectAcrRegistry(): Promise<AcrRegistryQuickPickItem> {
+        if (!(await this.azureAccount.waitForLogin())) {
+            await vscode.commands.executeCommand("azure-account.askForLogin");
+        }
+
+        const acrRegistryItem: AcrRegistryQuickPickItem = await vscode.window.showQuickPick(this.loadAcrRegistryItems(), { placeHolder: "Select Azure Container Registry", ignoreFocusOut: true });
+        return acrRegistryItem;
+    }
+
+    private async loadAcrRegistryItems(): Promise<AcrRegistryQuickPickItem[]> {
+        try {
+            await this.azureAccount.waitForFilters();
+            const registryPromises: Array<Promise<AcrRegistryQuickPickItem[]>> = [];
+            for (const azureSubscription of this.azureAccount.filters) {
+                const client = new ContainerRegistryManagementClient(
+                    azureSubscription.session.credentials,
+                    azureSubscription.subscription.subscriptionId!,
+                );
+
+                registryPromises.push(
+                    Utility.listAllAzureResource(client.registries, client.registries.list())
+                        .then((registries: Registry[]) => registries.map((registry: Registry) => {
+                            return new AcrRegistryQuickPickItem(registry, azureSubscription);
+                        })),
+                );
+            }
+
+            const registryItems: AcrRegistryQuickPickItem[] = ([] as AcrRegistryQuickPickItem[]).concat(...(await Promise.all(registryPromises)));
+            registryItems.sort((a, b) => a.label.localeCompare(b.label));
+            return registryItems;
+        } catch (error) {
+            error.message = `Error fetching registry list: ${error.message}`;
+            throw error;
+        }
+    }
+
+    private async selectAcrRepo(registryUrl: string, session: AzureSession): Promise<vscode.QuickPickItem> {
+        const acrRepoItem: vscode.QuickPickItem = await vscode.window.showQuickPick(this.loadAcrRepoItems(registryUrl, session), { placeHolder: "Select Repository", ignoreFocusOut: true });
+        return acrRepoItem;
+    }
+
+    private async loadAcrRepoItems(registryUrl: string, session: AzureSession): Promise<vscode.QuickPickItem[]> {
+        try {
+            const { accessToken, refreshToken } = await this.acquireToken(session);
+            this.refreshTokenArc = await this.acquireRefreshTokenArc(registryUrl, session.tenantId, refreshToken, accessToken);
```

Comment:
acr accesstoken will not in this response?

Reply:
Can we just use acquireAccessTokenAcr with the access token and refresh token? Besides, shall we cache the refresh token?

Reply:
No, not in the response.

Reply:
No. The access tokens for listing repos and tags are different in scope (compare Line 91 and Line 164). `refreshTokenArc` is already cached in the object property.